ydb/query/pool.py

```diff
@@ -0,0 +1,145 @@
+import time
+from typing import Callable
+
+from . import base
+from .session import (
+    QuerySessionSync,
+)
+from .. import issues
+from .._errors import check_retriable_error
+
+
+class RetrySettings(object):
+    def __init__(
+        self,
+        max_retries: int = 10,
+        max_session_acquire_timeout: int = None,
+        on_ydb_error_callback: Callable = None,
+        idempotent: bool = False,
+    ):
+        self.max_retries = max_retries
+        self.max_session_acquire_timeout = max_session_acquire_timeout
+        self.on_ydb_error_callback = (lambda e: None) if on_ydb_error_callback is None else on_ydb_error_callback
+        self.retry_not_found = True
+        self.idempotent = idempotent
+        self.retry_internal_error = True
+        self.unknown_error_handler = lambda e: None
+
+
+class YdbRetryOperationSleepOpt:
+    def __init__(self, timeout):
+        self.timeout = timeout
+
+    def __eq__(self, other):
+        return type(self) == type(other) and self.timeout == other.timeout
+
+    def __repr__(self):
+        return "YdbRetryOperationSleepOpt(%s)" % self.timeout
+
+
+class YdbRetryOperationFinalResult:
+    def __init__(self, result):
+        self.result = result
+        self.exc = None
+
+    def __eq__(self, other):
+        return type(self) == type(other) and self.result == other.result and self.exc == other.exc
+
+    def __repr__(self):
+        return "YdbRetryOperationFinalResult(%s, exc=%s)" % (self.result, self.exc)
+
+    def set_exception(self, exc):
+        self.exc = exc
+
+
+def retry_operation_impl(callee: Callable, retry_settings: RetrySettings = None, *args, **kwargs):
+    retry_settings = RetrySettings() if retry_settings is None else retry_settings
+    status = None
+
+    for attempt in range(retry_settings.max_retries + 1):
+        try:
+            result = YdbRetryOperationFinalResult(callee(*args, **kwargs))
+            yield result
+
+            if result.exc is not None:
+                raise result.exc
+
+        except issues.Error as e:
+            status = e
+            retry_settings.on_ydb_error_callback(e)
+
+            retriable_info = check_retriable_error(e, retry_settings, attempt)
+            if not retriable_info.is_retriable:
+                raise
+
+            skip_yield_error_types = [
```

Comment:
A code whould take the info from ErrorRetryInfo (sleep_timeout_seconds) or add additional field to the ErrorRetryInfo - for reuse same code for error logic.

Reply:
after discussion we've agreed  that this is out of scope